### ISSUE TITLE
Scrolling is not true 120hz and feels laggy on ProMotion iPhone 13

### ISSUE THREAD

**Issue by acoutts** [flutter/flutter#90675]
Even though the performance overlay says the frame timing should easily produce 120fps, it *feels* different than native iOS apps on the iPhone 13- a bit slower and janky. **this is not sksl jank**

I took some slow motion videos to demonstrate the scrolling performance issue. When you play it back, scrub frame by frame and you will see there is a lot more jank between screen updates on the flutter app vs the native messages one.

You'll see the performance profiler reporting 2.7ms/frame which in theory should mean it's capable of 370fps+ but as you can see there is a big difference between the frames on the native messages app and the flutter sample app.

Flutter app video (iPhone 13 Pro):
https://drive.google.com/file/d/1Qwqty1dCoGUyIfB5t5WeBf1s6czQhzFZ/view

Native iOS messages app (iPhone 13 Pro):
https://drive.google.com/file/d/1w5cw_PaIiMPTlyLuq2dKo86-KsWr-FEN/view

In case Google Drive applies some processing to the videos, here's a zip file with them untouched:
https://drive.google.com/file/d/1kqDEEI50Bs1uctgePvHKM-hJEE0QUhrA/view

## Steps to Reproduce

Example app available here:
https://github.com/acoutts/flutter_120hz_bug

**Expected results:** 
Flutter apps should run as smooth as native apps.

**Actual results:** 
Flutter performance does not match the native experience on 120hz devices.

<details>
  <summary>Logs</summary>

```
```

```
```

```
[✓] Flutter (Channel stable, 2.5.1, on macOS 11.6 20G165 darwin-x64, locale en-US)
    • Flutter version 2.5.1 at /Users/andrewcoutts/Projects/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffb2ecea52 (7 days ago), 2021-09-17 15:26:33 -0400
    • Engine revision b3af521a05
    • Dart version 2.14.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/andrewcoutts/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/andrewcoutts/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 13.0, Build version 13A233
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.1

[✓] VS Code (version 1.61.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.26.0

[✓] Connected device (4 available)
    • macOS (desktop)     • macos                     • darwin-x64     • macOS 11.6 20G165 darwin-x64
    • Chrome (web)        • chrome                    • web-javascript • Google Chrome 94.0.4606.61

• No issues found!
```

</details>

**Comment by acoutts** [comment]
This looks potentially related: https://twitter.com/ChristianSelig/status/1441419862417494018

>UIView.animateWithDuration APIs aren't clocked at 120Hz on iPhone 13? On UIScrollView, system ones, and Metal by the looks of it, rest is still 60Hz

**Comment by sroddy** [comment]
Flutter uses Metal, so it should be already capable of reaching 120hz. I think that the current issue is that the vsync waiter currently checks for the reported frame rate from the CADisplayLink that is still 60.
https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm#L103

The reason why this happens could be related to this:
https://www.macrumors.com/2021/09/24/apple-promotion-third-party-app-support/

I think it should be fairly simple to support the right refresh rate once it’s properly reported by the iOS SDK.

On the other hand I’m worried that supporting a completely adaptive scenario (dynamically switching between 10 and 120hz), like the one that will be natively supported in the upcoming future also by third-party apps, could be way less trivial to achieve

**Comment by acoutts** [comment]
Thanks  @sroddy

I just tried adding `CADisableMinimumFrameDurationOnPhone` to my Info.plist but it had no noticeable impact.

https://mobile.twitter.com/jamesthomson/status/1441461354435547138

**Comment by CodeGlitch** [comment]
According to apple:

The answer came from Apple today. The company told AppleInsider that the 120Hz ProMotion screen will work on all third-party apps although developers need to add a plist entry in order for their apps to work with the faster refresh rate.
https://www.phonearena.com/news/apple-says-120hz-refresh-rate-works-with-all-apps_id135301

**Comment by sroddy** [comment]
This is the official Apple documentation

https://developer.apple.com/documentation/quartzcore/optimizing_promotion_refresh_rates_for_iphone_13_pro_and_ipad_pro

**Comment by knopp** [comment]
It's not just the plist entry, it is also necessary to set `preferredFrameRateRange` on `CADisplayLink`. The range should ideally depend on what the application is doing, not just set to 120fps.

**Comment by Kounex** [comment]
For the time being, I showcased a sample implementation of how to enable the ProMotion feature: https://twitter.com/Kounexx/status/1442529256706396164

You can do this until there is an official implementation!

**Comment by chinmaygarde** [comment]
cc @jmagman

**Comment by Kounex** [comment]
The behaviour seems to be strange indeed. Looking at the documentation, it even suggests variations like this:

`CAFrameRateRange(minimum:8, maximum:15, preferred:0)`

Ultimately the best solution will probably be that Flutter applies different frame rates for different use cases (list scroll, expanding, progress bar etc.) instead of applying it in general. Testing the battery performance would be interesting as well!

**Comment by acoutts** [comment]
I think this needs to set in the engine, as per @sroddy:
>Flutter uses Metal, so it should be already capable of reaching 120hz. I think that the current issue is that the vsync waiter currently checks for the reported frame rate from the CADisplayLink that is still 60.
https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm#L103

Bigger issue might be we need an api exposed to change it at runtime.

**Comment by jmagman** [comment]
@chinmaygarde closed a similar issue a few years ago for the iPad Pro.  From https://github.com/flutter/flutter/issues/27259#issuecomment-521394675

> We already support 120Hz on devices with ProMotion displays. We get this out of the box because we don't set a [-[CADisplayLink preferredFramesPerSecond]](https://developer.apple.com/documentation/quartzcore/cadisplaylink/1648421-preferredframespersecond) on the vsync waiter implementation on iOS. If there is no preference, iOS sets this to the maximum supported display refresh rate. To verify, I captured this trace on an iPad with a ProMotion display. As marked, the frame interval is ~8.3ms (1/120).
> 
> <img alt="Screen Shot 2019-08-14 at 12 51 50 PM" width="842" src="https://user-images.githubusercontent.com/44085/63051833-d20bc480-be92-11e9-8522-5b581ee58314.png">

There's an open issue to add an iPad Pro to the devicelab for benchmarking.  https://github.com/flutter/flutter/issues/31450.

`preferredFramesPerSecond` is deprecated, maybe the problem is that this check should instead use `preferredFrameRateRange.preferred` on iOS 15+?  I would hope they would just be aliasing one to the other though.
https://github.com/flutter/engine/blob/adfea3cf1f6a7ac581a7717ef497e32a5bbb29f5/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm#L110

And `flutter create` should set `CADisableMinimumFrameDurationOnPhone` to the Info.plist for the iPhone 13 Pro.

**Comment by sroddy** [comment]
I think that the approach you are testing “kind-of” works because the screen refreshes at the max frame rate declared by any of the currently active animations.

if you add a CADisplayLink on the current run loop that refreshes at 120fps, all the CADisplayLink, including the one registered by the Flutter Engine, will start having their callback invoked at 120fps (unless they have a min value specified-which the one used by Flutter Engine is not having)

I think that there is no way to slow everything down below 60fps because the Flutter’s one is currently using the default settings and is set to run at 60fps min by default.

**Comment by acoutts** [comment]
@jmagman could we possibly bump the priority above p4? All native iOS apps support 120hz out of the box for things like scrolling, and it's only a potential issue when they have some kind of custom animation to show.

For flutter apps, everything including scrolling is being rendered at 60hz which makes flutter apps feel laggy by comparison and feels like a big regression in app performance for anybody who simply upgrades their phone. This affects everybody with a flutter app deployed in production right now.

**Comment by acoutts** [comment]
@sroddy it only sets to 60fps if it's not at least iOS 10.3. I used some print statements to see what a default CADisplayLink reports without setting `preferredFrameRateRange`.

<img width="1139" alt="Screen Shot 2021-09-27 at 6 19 54 PM" src="https://user-images.githubusercontent.com/9597207/134993088-5b873030-ec65-42f7-a62c-a9bcffbc2ed1.png">

The output:
```
2021-09-27 18:18:33.290238-0400 Runner[1979:597023] maximumFramesPerSecond before | 120
2021-09-27 18:18:33.292048-0400 Runner[1979:597023] display link preferredFrameRateRange before | CAFrameRateRange(minimum: 0.0, maximum: 0.0, __preferred: 0.0)
2021-09-27 18:18:33.292097-0400 Runner[1979:597023] display link preferredFrameRateRange after | CAFrameRateRange(minimum: 80.0, maximum: 120.0, __preferred: 120.0)
2021-09-27 18:18:33.294646-0400 Runner[1979:597023] maximumFramesPerSecond after | 120
```

Looking at how the flutter engine sets up the CADisplayLink, 

<img width="769" alt="Screen Shot 2021-09-27 at 6 20 40 PM" src="https://user-images.githubusercontent.com/9597207/134993164-41086110-f1d1-4996-a18b-17820d3230b7.png">

It appears that a CADispalyLInk by default returns 0 for everything, which means it would simply return `maximumFramesPerSecond`, which on my iPhone 13 Pro reports as 120 from the log output above.

This means flutter is already receiving the display refresh rate as 120.

**Comment by sroddy** [comment]
@jmagman @acoutts 

Depending on how the MetalKit view has been implemented inside the Engine (I didn't have the time to investigate) it may also be needed to play with the [preferredFramesPerSecond](https://developer.apple.com/documentation/metalkit/mtkview/1536027-preferredframespersecond) property of the MTKView itself.

This WWCD session is very informative about how to potentially deal with this "madness"
https://developer.apple.com/videos/play/wwdc2021/10147

But in order to really benefit from the ProMotion display features, I think devs would need a way to specify the desired frame rate of a given animation, so that the frames can be provided at the desired pace and the ProMotion display would automatically adapt its frame rate to optimise rendering and battery, given the various conditions.

**Comment by brucevang** [comment]
It's still showing 60hz after implementing the changes above. Testing a list view on my iPhone 13 Pro Max and it's still laggy compared to native apps and the App Store.

**Comment by alboiuvlad29** [comment]
Not sure wether this helps, but there seems to be a strange behaviour that could potentially be from Apple's side.

This is the content of `AppDelegate.swift` used in our app:
```
@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  var displayLink : CADisplayLink?
    
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    
    let controller = self.window.rootViewController as! FlutterViewController
    
    displayLink = CADisplayLink(target: self, selector: #selector(displayLinkCallback))
    displayLink!.add(to: .current, forMode: .default)
      if #available(iOS 15.0, *) {
          displayLink!.preferredFrameRateRange = CAFrameRateRange(minimum:120, maximum:120, preferred:120)
      } else {
          // Fallback on earlier versions
      }
      
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
    
    @objc func displayLinkCallback() {
         // let workingTime = displayLink!.targetTimestamp - CACurrentMediaTime()
        // At this moment in time, there are `workingTime` seconds available to
        // generate content for the next frame, so
        // Core Animation can render it to the display.
    }

}

```

with `<key>CADisableMinimumFrameDurationOnPhone</key><true/>`.

Starting the app for the first time on an iPhone 13 Pro Max, the app runs at `60Hz`.

Going to control centre and turning on `Low power mode`, then off will force the app to run in `120hz` mode until the app is force closed and started again. (Scrolling is as smooth as native, and transitions as well, etc).

I tried the same behaviour with `CADisableMinimumFrameDurationOnPhone` set to `false`, but it doesn't work and the app is locked at `60Hz`.

**Comment by k32net** [comment]
> I can confirm with that code in place and cycling low power mode, it runs at 120hz.

not working for me. neither on debug nor release mode.

**Comment by eseidelGoogle** [comment]
Thank you all so much for the engagement and suggestions.

My understanding from this bug is that although Flutter supports 120hz displays (see above comments), including ProMotion on the iPhone 13 Pro, we do not yet have the "turn on 120hz for scrolling or certain animations" heuristics built in to the Flutter framework.  So our current options are either "always 120hz" (which would burn battery) or "always 60hz" (iOS default).  @iskakaushik is in the middle of finishing up some dirty-region management work (to make all Flutter apps do less work when drawing) and then has indicated interest in tackling this.

I expect it will be a few weeks before we get to this.  Patches are otherwise welcome in the meanwhile.

**Comment by PedroAlvarado** [comment]
Is there a pseudo-official way to force "always 120hz"?

**Comment by maheshmnj** [comment]
>Is there a pseudo-official way to force "always 120hz"?

@PedroAlvarado you may try this [third-party plugin](https://pub.dev/packages/flutter_displaymode) and may also refer this discussion https://github.com/flutter/flutter/issues/35162

**Comment by iapicca** [comment]
> > Is there a pseudo-official way to force "always 120hz"?
> 
>  you may try this [third-party plugin](https://pub.dev/packages/flutter_displaymode) and may also refer this discussion #35162

@maheshmnj 
the package you referenced seems to be android only

**Comment by alexander-shustanov** [comment]
Hi! Are there any updates on this issue?

**Comment by jmagman** [comment]
We're actively discussing the best way to design this feature, see flutter.dev/go/variable-refresh-rate

**Comment by iskakaushik** [comment]
Hi All,

Thanks for the patience on this. Based on discussions with the team and as noted in [flutter.dev/go/variable-refresh-rate](https://flutter.dev/go/variable-refresh-rate), the next steps are the following:

1.  Engine will add support for rendering at the maximum supported refresh rate on a given iOS device. This should address the most common issue of native feel for scrolling and transitions.
2.  Required tooling/template changes to add the required preferences will be made in parallel. This change and the engine change will most likely reside behind a runtime flag.
3.  We will start measuring the impact of energy consumption and task scheduling (as we have smaller idle intervals at higher refresh rates) once that is possible. This is blocked on https://github.com/flutter/flutter/issues/92444.
4.  Based on what we learn from the measurements and field experiments, we will take action on changing the defaults.

**Comment by miguelpruivo** [comment]
Is there a reliable workaround as of now to force 120 Hz to be on all the time on these devices? (Despite bigger battery usage).

**Comment by cyanglaz** [comment]
Updates: this is being worked on https://github.com/flutter/engine/pull/29797, which is blocked by https://github.com/flutter/flutter/issues/85555

**Comment by Hixie** [comment]
here's a proposed api for the framework: https://github.com/flutter/flutter/compare/master...Hixie:freq

**Comment by tilemoon** [comment]
> here's a proposed api for the framework: https://github.com/flutter/flutter/compare/master...Hixie:freq

how to use the api. i clone your repo but not work for me.frame rate not changed.

**Comment by Hixie** [comment]
That branch is just a proposed API, it's not an implementation. It would need to be connected to back-end code in the engine to actually do something.

**Comment by pkoehler-altow-extern** [comment]
push

**Comment by 183839128** [comment]
Is there any alternative to modifying the engine 
2.8.0 still has this problem, is 2.8.1 fixed？

**Comment by acoutts** [comment]
Excited to see #96710 and flutter/engine#30900!

**Comment by prilepskiy** [comment]
Have some workaround here until the issue is resolved? My app is flickering with iPhone 13 Pro and 13 Pro Max and I get a lot of negative feedback from users.

I tried to play with this values in AppDelegate.swift `CAFrameRateRange(minimum:120, maximum:120, preferred:120)` from [here](https://github.com/flutter/flutter/issues/90675#issuecomment-930249845), but thad didn't help.

**Comment by hetao29** [comment]
I have the same issue, iphone13 pro + ios 15.3.

**Comment by jmagman** [comment]
The iOS 15 SDK is now available in the engine https://github.com/flutter/flutter/issues/85555.

**Comment by cyanglaz** [comment]
The engine fixed has been rolled into flutter/master.
If you would like to test this feature, please point your flutter to the master channel: 
```
$ flutter channel master
$ flutter upgrade
```
To enable 120hz. Also make sure you have `<key>CADisableMinimumFrameDurationOnPhone</key><true/>` in the  `<path_to_your_app>/ios/Runner/Info.plist`

**Comment by eric8810** [comment]
1. Does it resolve the laggy feel?
laggy feel of scroll and animations are gone, but its **still feels laggy when I use gesture navigation between routes**.

2. How does this impact battery life? Stress-testing your app with this change would be very helpful.
untest

3. You shouldn't have to manually update Info.plist, did the flutter tool add the key for you?
info.plist was updated with flutter create

**Comment by jmagman** [comment]
Thanks @eric8810

> still feels laggy when I use gesture navigation between routes

Can you file a new issue for this? Likely unrelated to the app refresh rate.

**Comment by felixwortmann** [comment]
Does it resolve the laggy feel?
- It mainly does. *If I scroll very slowly in a listview, it still does not seem to adapt perfectly so there is some jittering*.

How does this impact battery life? 
- I did not test this yet

You shouldn't have to manually update Info.plist, did the flutter tool add the key for you?
- Key was added using flutter create

**Comment by eric8810** [comment]
> Thanks @eric8810
> 
> > still feels laggy when I use gesture navigation between routes
> 
> Can you file a new issue for this? Likely unrelated to the app refresh rate.

Issues have been filed up, see #101653
